File: Java/5_Hash/LC_187_RepeatedDNASequences.java
class Solution {
    public List<String> findRepeatedDnaSequences(String s) {
        List<String> ans = new ArrayList<>();
        int n = s.length();
        Map<String, Integer> map = new HashMap<>();
        for (int i = 0; i + 10 <= n; i++) {
            String cur = s.substring(i, i + 10);
            int cnt = map.getOrDefault(cur, 0);
            if (cnt == 1)
                ans.add(cur);
            map.put(cur, cnt + 1);
        }
        return ans;
    }
}

public class LC_187_RepeatedDNASequences {

}

File: Swift/Leetcode.playground/Sources/5_Hash/LC_187_RepeatedDNASequences.swift
class Solution {
    func findRepeatedDnaSequences(_ s: String) -> [String] {
        var ans = [String]()
        let n = s.count
        if n < 10{
            return ans
        }
        var map = [String: Int64]()
        for i in 0...(n - 10) {
            let start = s.index(s.startIndex, offsetBy: i)
            let end = s.index(s.startIndex, offsetBy: i + 10)
            let cur = String(s[start..<end])
            let cnt = map[cur, default: 0]
            if cnt == 1 {
                ans.append(cur)
            }
            map[cur] = cnt + 1
        }
        return ans
    }
}